### PR TITLE
[nanogui] Update to 2020-06-04

### LINKED ISSUE #11743: [nanogui] update to 2020-04-05
Library name: NanoGUI

New version number: 2020-04-05

Other information that may be useful (release notes, etc...)

NanoGUI's repository has been moved [here](https://github.com/mitsuba-renderer/nanogui), and the [old one](https://github.com/wjakob/nanogui) is not going to be updated any more.

### PR DESCRIPTION
**Pull request by PhoebeHui**
Fixes https://github.com/microsoft/vcpkg/issues/11743

1. Update nanogui to new repo https://github.com/mitsuba-renderer/nanogui
2. Update to latest commits, and update the patch to match the source changes.
3. Disable to build examples for saving build time.

### CODE REVIEW

**Comment by PhoebeHui** [comment]
@BillyONeal, I notcied some PRs still failed on OSX, could you please have look?

Agent name: 'vcpkg-eg-mac-05'
Agent machine name: 'vcpkg-eg-mac-05'
Current agent version: '2.164.6'

Failures:
##[error]ENOENT: no such file or directory, stat '/Users/vagrant/Data/work/1/s/scripts/azure-pipelines/osx/initialize-environment.sh'

**Comment by BillyONeal** [comment]
@PhoebeHui if that happens, push a merge with origin/master to fix it. Pipelines appears to use old azure-pipelines.yml from before merge if GitHub hasn't remerged it itself

**Comment by PhoebeHui** [comment]
@BillyONeal, thanks for the information!

**Review by JackBoosY — APPROVED** [review]

File: ports/nanogui/CONTROL [hunk 1]
```diff
@@ -1,5 +1,6 @@
 Source: nanogui
-Version: 2019-09-23
-Homepage: https://github.com/wjakob/nanogui
+Version: 2020-06-04
+Homepage: https://github.com/mitsuba-renderer/nanogui
 Description: NanoGUI is a minimalistic cross-platform widget library for OpenGL 3.x or higher.
-Build-Depends: glfw3, nanovg, eigen3
\ No newline at end of file
+Build-Depends: glfw3, nanovg, eigen3
```
Comment (ras0219-msft) [inline ports/nanogui/CONTROL:5]: From the repo:
>The most visible change to developers is that it no longer relies on Eigen or Enoki ...

So I think we should be able to remove `eigen3` here?

File: ports/nanogui/fix-cmakelists.patch [hunk 2]
```diff
@@ -1,72 +1,100 @@
...
+   message(STATUS "Setting build type to 'Release' as none was specified.")
+   set(CMAKE_BUILD_TYPE Release CACHE STRING "Choose the type of build." FORCE)
+@@ -65,8 +59,8 @@ if (NOT NANOGUI_BACKEND)
+       CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64" OR
+       CMAKE_CXX_COMPILER MATCHES "/em\\+\\+(-[a-zA-Z0-9.])?$")
+     set(NANOGUI_BACKEND_DEFAULT "GLES 2")
+-  elseif (APPLE)
+-    set(NANOGUI_BACKEND_DEFAULT "Metal")
++  #elseif (APPLE)
++    #set(NANOGUI_BACKEND_DEFAULT "Metal")
```
Comment (ras0219-msft) [inline ports/nanogui/fix-cmakelists.patch:25]: We would prefer to use the Metal backend on apple if possible; can we look at fixing any issues another way?

File: ports/nanogui/fix-cmakelists.patch [hunk 3]
```diff
@@ -1,72 +1,100 @@
...
+-  ext/nanovg/src/nanovg.c
+-  ext/nanovg/src/stb_image.h
+   # Merge GLAD into the NanoGUI library (only if needed)
+   ${LIBNANOGUI_EXTRA_SOURCE}
+   # Fonts etc.
+@@ -504,6 +493,11 @@ add_library(nanogui-obj OBJECT
+   include/nanogui/nanogui.h
+ )
+ 
++find_path(EIGEN_INCLUDE_DIR Eigen/Core)
```
Comment (ras0219-msft) [inline ports/nanogui/fix-cmakelists.patch:58]: This should no longer be needed.

File: ports/nanogui/fix-cmakelists.patch [hunk 4]
```diff
@@ -1,72 +1,100 @@
...
+ endif()
+ 
+-if (NANOGUI_BUILD_GLFW)
+-  set(NANOGUI_LIB_CONTENTS ${NANOGUI_LIB_CONTENTS} $<TARGET_OBJECTS:glfw_objects>)
+-endif()
+-
+ add_library(nanogui ${NANOGUI_LIBRARY_TYPE} ${NANOGUI_LIB_CONTENTS})
+ 
++find_package(nanovg CONFIG REQUIRED)
++find_package(Eigen3 CONFIG REQUIRED)
```
Comment (ras0219-msft) [inline ports/nanogui/fix-cmakelists.patch:77]: Should no longer be needed.

File: ports/nanogui/fix-cmakelists.patch [hunk 5]
```diff
@@ -1,72 +1,100 @@
...
++find_package(glfw3 CONFIG REQUIRED)
++set(NANOGUI_EXTRA_LIBS ${NANOGUI_EXTRA_LIBS} glfw nanovg::nanovg Eigen3::Eigen)
```
Comment (ras0219-msft) [inline ports/nanogui/fix-cmakelists.patch:79]: `Eigen3::Eigen` here should no longer be needed.

**Review by ras0219-msft — CHANGES_REQUESTED** [review]

File: ports/nanogui/CONTROL (hunk 1 above)
Comment (PhoebeHui) [reply ports/nanogui/CONTROL]: Removed, thanks!

File: ports/nanogui/fix-cmakelists.patch (hunk 3 above)
Comment (PhoebeHui) [reply ports/nanogui/fix-cmakelists.patch]: Removed, thanks!

File: ports/nanogui/fix-cmakelists.patch (hunk 4 above)
Comment (PhoebeHui) [reply ports/nanogui/fix-cmakelists.patch]: Removed, thanks!

File: ports/nanogui/fix-cmakelists.patch (hunk 5 above)
Comment (PhoebeHui) [reply ports/nanogui/fix-cmakelists.patch]: Removed, thanks!

File: ports/nanogui/fix-cmakelists.patch (hunk 2 above)
Comment (PhoebeHui) [reply ports/nanogui/fix-cmakelists.patch]: To support Metal, we need to add new port 'nanovg_metal' firstly, it has wjakob/nanovg_metal which used in this project and official repo ollix/MetalNanoVG. I will try the official one.

File: ports/nanogui/portfile.cmake [hunk 6]
```diff
@@ -1,26 +1,23 @@
-include(vcpkg_common_functions)
-
-if(VCPKG_TARGET_IS_UWP)
-    message(FATAL_ERROR "nanogui doesn't support UWP.")
-endif()
+vcpkg_fail_port_install(ON_ARCH "arm" "arm64" ON_TARGET "uwp" )
```
Comment (BillyONeal) [inline ports/nanogui/portfile.cmake:1]: This appears to block arm64 desktop and not block x86-uwp?

**Comment by BillyONeal** [comment]
@PhoebeHui are you still working on this one?

**Comment by PhoebeHui** [comment]
We need add 3 new ports for supporting this changes, I will reopen this PR when I have time.

**Comment by Tanishq-Banyal** [comment]
> We need add 3 new ports for supporting this changes, I will reopen this PR and continue the work when I have time.

You can add Apple support later on, please add windows support for now

**Pull request closed without merging**